File: exercism/elixir/grade-school/school.exs
defmodule School do
  @moduledoc """
  Simulate students in a school.

  Each student is in a grade.
  """

  @doc """
  Add a student to a particular grade in school.
  """
  @spec add(map, String.t, integer) :: map
  def add(db, name, grade) do
    db |> Map.merge(%{ grade => [name] }, fn _k, curr, acc -> acc ++ curr end)
  end

  @doc """
  Return the names of the students in a particular grade.
  """
  @spec grade(map, integer) :: [String.t]
  def grade(db, grade) do
    db |> Map.get(grade, [])
  end

  @doc """
  Sorts the school by grade and name.
  """
  @spec sort(map) :: [{integer, [String.t]}]
  def sort(db) do
    db
    |> Map.to_list
    |> Enum.sort(fn {a, _}, {c, _} -> a < c end)
    |> Enum.map(fn {k, ls} -> {k, ls |> Enum.sort(fn a, b -> a < b end)} end)
  end
end

File: exercism/haskell/grade-school/src/School.hs
module School (School, add, empty, grade, sorted) where

import qualified Data.Map as M
import qualified Data.List as L

type Grade = Int
type Student = String
type Students = [Student]
type School = M.Map Int Students

add :: Grade -> String -> School -> School
add n name db = M.unionWith (++) db $ M.fromList [(n, [name])]

empty :: School
empty = M.empty

grade :: Grade -> School -> Students
grade g db = case M.lookup g db of
  Just ls -> ls
  Nothing -> []

sorted :: School -> [(Grade, Students)]
sorted m = fmap sortStudents schoolLs
  where
    schoolLs = M.toList m
    sortStudents (n, ls) = (n, L.sort ls)
